streaming/base/dataloader.py

```diff
@@ -0,0 +1,21 @@
+# Copyright 2022 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Streaming DataLoader."""
+
+from typing import Any, Dict
+
+from torch.utils.data import DataLoader
+
+from streaming.base.dataset import Dataset
+
+class StreamingDataLoader(DataLoader):
+
+    def state_dict(self):
+        if isinstance(self.dataset, Dataset):
+            return self.dataset.state_dict(self._iterator._num_yielded)
```

Comment:
Maybe it’s safer for future changes to do a pass-through __iter__ in this class and track num_yields ourself?

Reply:
What do you mean by pass-through iter? Like, wrap the iterator? 
`self._iterator = StreamingIterator(self._iterator)`?

Reply:
Something like…?

```
for batch in super().__iter__():
    self.num_yielded += len(batch[0])
```

But then I remembered dataloaders are only managing their set of workers, so this needs dist right?

Reply:
hm... I'm not sure if this would work. You need to override the actual iterator instead of just the dataloader I think? it might interplay in a messy way otherwise. Would need to try it out. 

I think it should be fine without dist. IIRC dataloader runs on main thread, not per worker, and each rank surfaces a batch. We need to multiply by # of ranks for batch size, but that should be fine?

Reply:
Sorry, meant avoiding cracking open DataLoader where possible, for example

```
class DataLoader(object):
    def __init__(self, num_batches, batch_size):
        self.num_batches = num_batches
        self.batch_size = batch_size

    def __iter__(self):
        self._num_yielded = 0
        for i in range(self.num_batches):
            a = i * self.batch_size
            yield list(range(a, a + self.batch_size))
            self._num_yielded += self.batch_size


loader = DataLoader(10, 4)
for batch in loader:
    print(batch)
print(loader._num_yielded)


class StreamingDataLoader(DataLoader):
    def __iter__(self):
        self.num_yielded = 0
        for batch in super().__iter__():
            yield batch
            self.num_yielded += len(batch)


loader = StreamingDataLoader(10, 4)
for batch in loader:
    print(batch)
print(loader.num_yielded)
```